File: test-case/tree/mirror/mirror.pl
use strict;
use warnings;
use List::Util qw(min max);
use Soda::DS::PriorityQueue;
use Soda::Leetcode::ListNode;
use Soda::Leetcode::NestedInteger;
use Soda::Leetcode::TreeNode;

*ListNode:: = \*Soda::Leetcode::ListNode::;
*NestedInteger:: = \*Soda::Leetcode::NestedInteger::;
*PriorityQueue:: = \*Soda::DS::PriorityQueue::;
*TreeNode:: = \*Soda::Unittest::TestWork::;

# step [1]: implement solution function
# @param {TreeNode} root
# @return {TreeNode}
sub mirror {
    my $root = shift;
    return undef unless defined $root;
    mirror($root->left);
    mirror($root->right);
    my $temp = $root->left;
    $root->set_left($root->right);
    $root->set_right($temp);
    $root;
}

package main;
use Soda::Unittest::Utils qw(from_stdin);
use Soda::Unittest::TestWork qw(create for_struct);
use Soda::Unittest::Validators qw(for_array for_array2d);
# step [2]: setup function/return/arguments
my $work = create(\&mirror);
# my $work = for_struct(PACKAGE::);
# $work->validator(sub { $_[0] vs $_[1]; return boolean });
$work->compare_serial(1);
print $work->run(from_stdin());

File: test-case/tree/mirror/mirror.js
const {ListNode, NestedInteger, TreeNode} = require('soda/leetcode');
const {TestWork, Utils, Validators} = require('soda/unittest');

/**
 * @param {TreeNode} root
 * @return {TreeNode}
 */
var mirror = function(root) {
    if (!root) {
        return null;
    }
    mirror(root.left);
    mirror(root.right);
    [root.left, root.right] = [root.right, root.left];
    return root;
};

const work = TestWork.create(mirror);
// work.validator = (x, y) => { ... };
work.compareSerial = true;
console.log(work.run(Utils.fromStdin()));

